src/backend/gporca/libgpopt/src/base/CDistributionSpecHashed.cpp

```diff
@@ -510,6 +524,8 @@ CDistributionSpecHashed::FMatchHashedDistribution
 		}
 		if (!fSuccess)
 		{
+			if (equiv_distribution_exprs == NULL && pdshashed->FDistributionSpecHashedOnlyOnGpSegmentId())
```

Comment:
Shouldn't we check both this and pdshashed for distribution on gp_segment_id?
```suggestion
			if (0 == ul && FDistributionSpecHashedOnlyOnGpSegmentId() && pdshashed->FDistributionSpecHashedOnlyOnGpSegmentId())
```

Reply:
Since you're pointing this out, I'm starting to feel current approach I took to handle group by is too much of a hack.  Right now pdbhashed is the distribution based on predicate in the query and this distribution is the actual distribution policy of the table (i.e. it won't be gp_segment_id).  I'll poke around to see if I can come up with something better.

Reply:
I tried another approach hoping to achieve the symmetry you mentioned.

I tried including GpSegmentId as a distribution column in RetrieveRelDistributionCols().  But that seemed to cascade a bit as then I needed to add distribution columns to the grouping columns side of distribution spec to match properly.

Maybe there's a better approach, but from exploring a bit, I feel like a little less sad about the current implementation.